src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -132,6 +158,211 @@ internal void FinalizeDebugInfo()
+                QsNullable<QsLocation> debugLoc = spec.Location;
+                if (debugLoc.IsNull)
+                {
+                    return this.CreateFunctionNoDebug(mangledName, signature);
+                }
+
+                Position absolutePosition = debugLoc.Item.Offset; // Because spec is its own namespace, the position is already an absolute position
+
+                if (retDebugType != null && !argDebugTypes.Contains(null))
+                {
+                    // create the function with debug info
+                    DebugFunctionType debugSignature = new DebugFunctionType(signature, curDIBuilder, DebugInfoFlags.None, retDebugType, argDebugTypes!);
+                    IrFunction func = this.Module.CreateFunction(
+                        scope: curDIBuilder.GetCompileUnit(),
+                        name: shortName,
+                        mangledName: mangledName,
+                        file: debugFile,
+                        linePosition: ConvertToDebugPosition(absolutePosition),
+                        signature: debugSignature,
+                        isLocalToUnit: true, // we're using the compile unit from the source file this was declared in
+                        isDefinition: isDefinition,
+                        scopeLinePosition: ConvertToDebugPosition(absolutePosition), // TODO: Could make more exact bc of formatting and white space (see lastParamLocation in Kaleidescope tutorial)
+                        debugFlags: DebugInfoFlags.None,
+                        isOptimized: false,
+                        curDIBuilder);
+
+                    this.EmitLocation(absolutePosition, func.DISubProgram);
+                    return func;
+                }
+                else
+                {
+                    return this.CreateFunctionNoDebug(mangledName, signature);
+                }
+            }
+            else
+            {
+                return this.CreateFunctionNoDebug(mangledName, signature);
+            }
+        }
+
+        /// <summary>
+        /// Creates the debug info for a function argument.
+        /// </summary>
+        /// <param name="spec">The function's <see cref="QsSpecialization"/></param>
+        /// <param name="name">The name of the function's argument.</param>
+        /// <param name="value">The value of the function's argument.</param>
+        /// <param name="argNo">The index of the argument (expected to be 1-indexed).</param>
+        internal void CreateFunctionArgument(QsSpecialization spec, string name, IValue value, int argNo)
+        {
+            if (this.DebugFlag)
+            {
+                // get the DIBuilder
+                string sourcePath = spec.Source.CodeFile;
+                DebugInfoBuilder curDIBuilder = this.GetOrCreateDIBuilder(sourcePath);
+                DIFile debugFile = curDIBuilder.CreateFile(Path.ChangeExtension(sourcePath, ".c")); // TODO: change this now that we have a language solution
```

Comment:
I'd suggest to address this todo. As far as I know, things work correctly on Linux, and the remaining Windows issues are also reproduceable when debugging C++, correct?

Reply:
The Windows issues aren't reproduceable when debugging C++ because it only comes up when we use a language enum from outside the C family. Things work correctly on Linux if I hand-edit the .c extension in the LLVM IR to the .qs extension. However, I discovered yesterday that when I automate outputting the .qs extension into the LLVM IR file, I'm hitting an LLVM bug on both Windows and Linux (because of some side effects in the DebugInfoBuilders). I filed a bug with LLVM and think I have a workaround. I'll update with that once I verify the workaround.